### DISCUSSION TITLE
About building an if-else function in yalmip

### DISCUSSION THREAD

**Discussion by 6qoq9** [yalmip/YALMIP#1197] [Problems using YALMIP]
Hello everyone

I want to build a loop function on the decision variable, as shown in the figure below, but there is an error using the implies function in the yalimp solver.
![微信截图_20221121194819](https://user-images.githubusercontent.com/98142766/203052699-068316c7-924d-4a4f-83a6-bafc989acac4.png)

here is my model's  constraints for your reference.
clc;clear;
T=96;
Pnet=sdpvar(T,1,'full');
Pgrid=sdpvar(T,1,'full');
Temp_net=binvar(T,1,'full');

load case1.mat
Constraints=[];
for k=1:T
Constraints = [Constraints, implies(Temp_net(k,1),
[Pnet(k,1)>=0,Price(k,1)>Peakprice,Pgrid(k,1)==Pnet(k,1)])];%If Pnet>0, and Price>peakprice, Pgrid==Pnet(sell excess power to main grid)

end
Program error
One of the constraints evaluates to a FALSE LOGICAL variable. Your model is infeasible
Constraints = [Constraints, implies(Temp_net(k,1),
[Pnet(k,1)>=0,Price(k,1)>Peakprice,Pgrid(k,1)==Pnet(k,1)])]

**Comment by 6qoq9** [comment]
For functions with multiple nested loops, can you use the implies function? For example implies(Temp_net(k,1), [Pnet(k,1)>=0,Price(k,1)>Peakprice,Pgrid(k,1)==Pnet(k,1)])
Among them, Pnet(k,1)>=0, Price(k,1)>Peakprice are the judgment conditions, and Pgrid(k,1)==Pnet(k,1) is the assignment result

**Comment by liujin1321** [comment]
I have also encountered this problem, may I ask if you have solved this problem? If so, can I ask you how you solved it?

**Reply by johanlofberg** [reply]
You would have to define what "this problem" is (and preferably you should not piggy-back old threads but open your own with a detailed minimal description of your particular problem.)

**Reply by liujin1321** [reply]
Thank you for your reply! 
For the individual if constrain, I read your release and learn that you can do it with the implies statement. But what about the nested if the constraint can be expressed with implies?
The example is:

![image](https://user-images.githubusercontent.com/117717679/224269193-cefad1d4-2e32-460b-a529-a947c55c42cb.png)

**Reply by johanlofberg** [reply]
You just work with the different cases you have in total, here you have four different cases, [Pnet>=0, whathappens], [Pnet<=0, price==high,whathappens], [Pnet<=0, price=mid,whathappens] and [Pnet<=0, price==low,whathappens]

**Reply by liujin1321** [reply]
Thank you for your reply!
I set up the constraint according to your suggestion, but there is such an error, may I ask how to solve this?

![image](https://user-images.githubusercontent.com/117717679/224478088-33a74edc-e41c-43f8-9a58-4143839b38cb.png)
![image](https://user-images.githubusercontent.com/117717679/224478113-705ee870-27ae-4c6b-9be3-a47106b5e8fc.png)

**Reply by johanlofberg** [reply]
My guess is that e_price is a constant, and it makes no sense to include a statement which either is known to be false (or true)

REasonable code is
if e_price ==...
  add these constraintse
else
  add these constraintse
end

Anyhow, you basically added a constant constraint of the type 1==0 to the model somewhere, so just look at the stuff you include in the constraint to see which evaluates to a constant instead of a constraint

**Reply by liujin1321** [reply]
Thank you for your advice, I have successfully solved this problem! Thank you very much!

And I have another question.:
When I set: f=(Pshift+pfix)./pv, it can run successfully.
However, when I set f=(Pshift+pfix)./(pv+BT_dis), here comes an error .

![image](https://user-images.githubusercontent.com/117717679/224495615-af7c2f78-05e2-459d-919f-4736c7cda044.png)

![image](https://user-images.githubusercontent.com/117717679/224495594-019a1223-44ec-4f99-b5d8-e3419cdc059d.png)

I hope to get your reply!

**Reply by johanlofberg** [reply]
That message reveals two things which both are problematic

1. f is a vector. That makes no sense. It could be that already the case which works for you is a vector too, but yalmip then solves the problem for every element in that vector, which probably isn't what you want, and the solution you looked at is just the solution corresponding to the last element

2. You are dividing by a decision variable, i.e. BT_dis is a decision variable. That leads to  model which you cannot solve using your solver (even if you fixed the issue with f being a vector, i.e. some missing sum or what ever it is you have forgotten.). You basically have to come up with a model where you don't require this division. With gurobi, you can add new variables and replace `x/y` with `x*z` with a new constraint `y*z == 1` but you have to realize that you get a very hard nonconvex quadratic model

**Reply by liujin1321** [reply]
Thank you very much for your answer! I will modify it according to your opinion, and express my thanks again!

**Reply by liujin1321** [reply]
Dear Johan,
I'm sorry to trouble you again. But I do have a problem.
I would like to ask if we can implement the nesting of two conditional statements using implicit syntax. For example:
![image](https://github.com/yalmip/YALMIP/assets/117717679/8df5ba68-b645-4a20-81b7-bdc13a067390)
A, B, C and D are all time-dependent variables.

I can successfully create a condition statement, but for the nested condition statement, I do not know how to do it using implies.
Looking forward to your reply！
